### PR TITLE
Implemented PRs from original repo - EXTERNAL CONTRIBUTOR

### PR DESCRIPTION
**Pull request by Stefano-Gasco**
# Description
Implementation (with slight modifications) of the PR proposed here in the [oasis-open repo](https://github.com/oasis-open/cti-taxii-server/pull/175/files#diff-ef2cef9f88b4fe09ca3082140e67f5ad34fb65fb6e228f119d3812261ae51449). The [user](https://github.com/chisholm) proposing the changes has few contributions over the last year, mostly on CTI open repos.
The implementation in the PR allegedly improves the memory and mongoDB processes, which we think is the problem we're facing when pushing identities to the server. Additionally, improves several smaller processes, based on previous PR never accepted. We've already implemented some of the changes.

# Review process
The review process was quite fast (~3h). The scope of the review was to assess the quality of the code, understand the logic behind the changes and check for unexpected connections, encrypted code blocks and weird behaviors in general.
**It's important to note that more than 7k lines have been modified/added/removed**, and an in-depth code review, assessing the impact of the changes, possible side effects and functional and non functional behaviors will take a lot of time, tools and resources; **The current code review assess that there aren't obvious problems.**

# Review outcome

- The code quality has been drastically improved, increasing readability, decreasing coupling and, for the most part, aligning with industry standard on code quality
- There are no obvious errors/bugs in the code; the functionality has not been tested
- There aren't obvious malign code blocks, and there aren't encrypted parts. The proposed implementation seems genuine.

# Notes to reviewers
**Please take care on reviewing the code!** After personal reviews, pair reviews could be done as an additional check. Additional ways of validating the code, other than simple code review, are welcome.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -113,23 +113,26 @@ The <config_file> contains:
...
 To use the Mongo DB back-end plug, include the following in the <config-file>:
 
-.. code-block:: json
+.. code-block:: text
 
     {
          "backend": {
             "module_class": "MongoBackend",
-            "uri": "<Mongo DB server url>  # e.g., 'mongodb://localhost:27017/'"
+            "uri": <Mongo DB server url>  # e.g., 'mongodb://localhost:27017/'
```
Comment (gtsaraias) [inline README.rst:133]: Do we need to add new lines in this config?

File: medallion/common.py [hunk 2]
```diff
@@ -141,34 +86,150 @@ def datetime_to_string_stix(dttm):
...
 
 def float_to_datetime(timestamp_float):
     """Given a floating-point number, produce a datetime instance"""
-    return dt.datetime.utcfromtimestamp(timestamp_float)
+    result = dt.datetime.utcfromtimestamp(timestamp_float)
+    result = result.replace(tzinfo=dt.timezone.utc)
+    return result
 
 
 def string_to_datetime(timestamp_string):
```
Comment (gtsaraias) [inline medallion/common.py:99]: Seems to revert our TZ fix.

File: README.rst (hunk 1 above)
Comment (Stefano-Gasco) [reply README.rst:133]: It's not actually a new line. I think that, for some reason (?), he decided to put everything in double quotes

File: medallion/common.py (hunk 2 above)
Comment (Stefano-Gasco) [reply medallion/common.py:99]: Yes, but he's reworking the timezone handling in general; maybe we should accept and eventually reintegrate the change?

**Review by gtsaraias — APPROVED** [review]

**Pull request merged**